### ISSUE TITLE
Option to skip Validate() methods generation is gone

### ISSUE THREAD

**Issue by kolyshkin** [go-swagger/go-swagger#2077]
## Executive summary

We use swagger to generate model and operation API for Docker Engine (aka [Moby](https://github.com/moby/moby)), and we believe that types and structures should be separated from the code. This is why we are using an option to disable generating `Validate()` methods.

Unfortunately, this option has disappeared, making us stick to a very old go-swagger codebase, which is becoming more and more problematic, e.g.
* problems with unstable/unsorted output (https://github.com/go-swagger/go-swagger/issues/846, example: https://github.com/moby/moby/pull/38832#issuecomment-472106785)
* golang 1.13 compatibility issue (https://github.com/go-swagger/go-swagger/issues/2058)

## Details

Both `swagger generate model` and `swagger generate operation` used to have an option to disable generating `Validate()` methods. The option name was `--skip-validator` (what was in line with other options such as `--skip-struct` or `--skip-responses`), it was added:

* for `generate model`: by commit 38dbdd808fb157f84724e19269d310128f7df71d ("move generator2 to generator package", Feb 16 2015, possibly earlier but there's no git history AFAICS)
* for `generate operation`: by commit 041580674330b64a393d7353ca13ffe4ab540924 ("Add a --skip-validator flag for generate operation", Oct 5 2016)

Unfortunately, this option was silently removed from both commands by commit d2aae8150c526d861828ee0e8a16623df01d3fb5 ("Added an option in genOpts", Jan 4, 2018).

A similarly named (but with very different semantics) option, `--skip-validation`, was added by commit 78f848e7f01c8dca5c2744c69b08d9732d60b537 ("validate spec prior to gen. client/server", Jan 8, 2017). As suggested in https://github.com/go-swagger/go-swagger/commit/d2aae8150c526d861828ee0e8a16623df01d3fb5#commitcomment-29777292, a possible user confusion between `--skip-validation` and `--skip-validator` was the reason for removal, although it's not very clear.

We can obviously work around that by forking go-swagger and applying fixes on top of the old code, but I believe there's a better way.

## Proposed solution

Reinstate the removed functionality.

To the best of my understanding, the biggest problem here is to avoid user confusion between `--skip-validation` and `--skip-validator`. There are many ways to solve it; I propose to:
 * add `--generate` option, accepting a `string:bool` type, for example: `--generate struct:false --generate validator:false`.
* for backward compatibility, add the following hidden options
`--skip-struct` (equivalent of `--generate struct:false`)
`--skip-handler`
`--skip-parameters`
`--skip-responses`
`--skip-url-builder`

**Comment by kolyshkin** [comment]
@fredbi @diego-fu-hs @casualjim PTAL

I am ready to work on the above proposal, as soon as we agree on the way forward. Please let me know what you think

**Comment by casualjim** [comment]
@kolyshkin we allow for fully customizing the generation process. This is a very similar use case as what you've laid out with your 2 options. All that is required is creating a yaml config file.

In addition to just skipping bits and pieces it allows you to add extra templates or remove some templates.
the docs with some examples are here: https://goswagger.io/use/template_layout.html

Personally I agree with the feature, but I wonder if the config file isn't more appropriate.  Irrespective of what you chose to do I look forward to the PR

**Comment by fredbi** [comment]
@casualjim well it occurs that this generated piece pertains to so called "protected templates". To be fair, this part is not completely customizable with go-swagger _as is_. Maybe we should just drop this protected thing altogether.
@kolyshkin  I believe the original filter with schema generation templates has remained in place, only the CLI option has been dropped (and mapping of it to the generation model). However, many changes have been added since this disappeared, so this probably will require a little tidying up of existing templates.

**Comment by fredbi** [comment]
@kolyshkin it is not clear to me what `--skip-struct` would bring compared to the existing `--skip-models`.

**Comment by casualjim** [comment]
@fredbi protected only applies if you don't use a custom yaml config.

**Comment by fredbi** [comment]
@casualjm oh really? I missed that part... Thanks for rectifying my false assertion. There is always something to be learnt on this codebase ... :)

**Comment by vuongabc92** [comment]
Hello everyone,

I'm sorry because this is an old post but is there any solution for this one. 
I really want to remove those 4 methods when generating model (Validate, validateMessage, MarshalBinary, UnmarshalBinary) I want to keep the model clean with just struct (using option skip-exists then manually remove unwanted method but I think it is a bad solution).
I'm very new to go swagger and I have researched a lot but I still stuck at this.

Please share if you guys have any solution @kolyshkin , @casualjim , @fredbi .
Really sorry for disturbing you.

Thank you very much.

**Comment by fredbi** [comment]
@vuongabc92  At the moment, you can achieve this by customizing the template.

Prepare a custom codegen config like explained here https://goswagger.io/use/template_layout.html#server-generation, then replace model generation by your own template. You may copy the [original](https://github.com/go-swagger/go-swagger/blob/master/generator/templates/schemavalidator.gotmpl) and remove the code you don't want then place in the location defined by your custom config.

We removed this option from the cli quite some time ago, because a lot of people got confused between `spec validation` and `generate validator`.

**Comment by casualjim** [comment]
But you have to know that a go-swagger generated server won't function without those methods.
And if  you're not using the go-swagger server, then maybe `swagger generate model` is more appropriate for your use case

**Comment by vuongabc92** [comment]
@fredbi @casualjim It works just fine.
Thank you very much for the support and sorry for the noise.

**Comment by YanzhaoLi** [comment]
@casualjim  
So all the model used by go-swagger server should have these methods including `Validate()`?  Thus, the imported model (by `x-go-import`) without these method won't work. We have to generate duplicate model by go-swagger and translated to the real object that we want import.

#2285

**Comment by casualjim** [comment]
That is correct, but you can make your own templates. The validate method is required because our handler templates need to validate the input

Alternatively you could implement the validate method on your own models, if you don't care about validation you can always leave the method empty. And you can use type definition or embedding strategy if you already have a `Validate` method and they would clash

**Comment by YanzhaoLi** [comment]
@casualjim   Thanks. 
BTW, Can I make a  template that only affects specific objects, while the other objects still use the default handler templates?

**Comment by ysomad** [comment]
Any updates on this?

**Comment by Carseason** [comment]
Too many Validate methods are not friendly to embedded devices
Are there any updates on this?
